### PR TITLE
Optimize SG generation

### PR DESCRIPTION
**Pull request by YairSlobodin1**

### CODE REVIEW

File: pkg/synth/sg.go [hunk 1]
```diff
@@ -6,97 +6,142 @@ SPDX-License-Identifier: Apache-2.0
...
+		sgSrc := s.Result.LookupOrCreate(sgSrcName)
+		sgSrc.Attached = []ir.ID{ir.ID(sgSrcName)}
+		rule := &ir.SGRule{
+			Remote:      sgRemote(&s.Spec.Defs, dstEndpoint),
+			Direction:   ir.Outbound,
+			Protocol:    trackedProtocol.Protocol,
+			Explanation: reason,
+		}
+		sgSrc.Add(rule)
+	}
```
Comment (zivnevo) [inline pkg/synth/sg.go:84]:
```suggestion
	internalSrc, _, internal := internalConn(conn)
	if !internalSrc {
		return
	}
	reason := explanation{internal: internal, connectionOrigin: conn.Origin, protocolOrigin: trackedProtocol.Origin}.String()
	sgSrcName := ir.SGName(srcEndpoint.Name)
	sgSrc := s.Result.LookupOrCreate(sgSrcName)
	sgSrc.Attached = []ir.ID{ir.ID(sgSrcName)}
	rule := &ir.SGRule{
		Remote:      sgRemote(&s.Spec.Defs, dstEndpoint),
		Direction:   ir.Outbound,
		Protocol:    trackedProtocol.Protocol,
		Explanation: reason,
	}
	sgSrc.Add(rule)
```

File: pkg/synth/sg.go [hunk 2]
```diff
@@ -6,97 +6,142 @@ SPDX-License-Identifier: Apache-2.0
...
+func (s *SGSynthesizer) allowConnectionToDst(conn *ir.Connection, trackedProtocol ir.TrackedProtocol,
+	srcEndpoint, dstEndpoint *namedAddrs) {
+	_, internalDst, internal := internalConn(conn)
+	reason := explanation{
+		internal:         internal,
+		connectionOrigin: conn.Origin,
+		protocolOrigin:   trackedProtocol.Origin,
+	}.String()
+
+	if internalDst {
```
Comment (zivnevo) [inline pkg/synth/sg.go:97]: Same here

File: cmd/subcmds/sg.go [hunk 3]
```diff
@@ -24,8 +25,8 @@ func NewSGCommand(args *inArgs) *cobra.Command {
 			if err != nil {
 				return err
 			}
-			collection := synth.MakeSG(spec)
-			return writeOutput(args, collection, &spec.Defs.ConfigDefs)
+			sgSynthesizer := synth.SGSynthesizer{Spec: spec, Result: ir.NewSGCollection()}
```
Comment (zivnevo) [inline cmd/subcmds/sg.go:28]: Might be nicer to provide a `NewSGSynthesizer(spec *ir.Spec)` method.
This allows to un-export the internal `Spec` and `Result` members, and doesn't require the caller to know how to initialize them.

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: cmd/subcmds/sg.go (hunk 3 above)
Comment (zivnevo) [reply cmd/subcmds/sg.go]: How about having a common interface for both synthesizers?
```golang
type Synthesizer interface {
	Synth() ir.Collection
}
```
Then you can have a single function to implement common code in the `RunE` of both `sg` and `acl` sub-commands.

Comment (YairSlobodin1) [reply cmd/subcmds/sg.go]: done

**Review by zivnevo — APPROVED** [review]

**Pull request merged**